### PR TITLE
Basic health management

### PR DESCRIPTION
**Pull request by ghost**
Implements #73.

(Github now also has squash on merge I believe. If you want me to squash on my branch anyhow just tell me.)

### CODE REVIEW

File: src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java [hunk 1]
```diff
@@ -48,39 +52,105 @@
...
+                player.setMaxHealth(newMaximumHealth);
+            }
+        }
+
+        if (isSettingCurrentHealth) {
+            double newCurrentHealth = setCurrentHealthSpec.value(optionSet);
+            for (Player player : players) {
+                double playerMaxHealth = player.getMaxHealth();
+                // Spigot 1.9+ throws an error trying to set a player's health higher than his maximum health
+                player.setHealth(newCurrentHealth > playerMaxHealth ? playerMaxHealth : newCurrentHealth);
```
Comment (Eluinhost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java]: `Math.min(newCurrentHealth, player.getMaxHealth())`?

**Comment by Eluinhost** [comment]
If the client has potential to crash at a particular amount it'd be logical to hardcode the limit. If it's an unknown value then it should just give a warning when going above a certain preset amount. I don't really see the merit of adding a config for the 'max max' setting, it'll just restrict people. Maybe change the config to have a 'show warnings on high max health values or something: true/false' instead?

If we're going to change the /h permission to `.health` we'll have to remember to include it in the changelog so permissions can be updated.

File: src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java [hunk 2]
```diff
@@ -48,39 +52,105 @@
...
+                        .setType("New player health")
+                );
+
+        setMaximumHealthSpec = parser
+                .accepts("setMaximum", "Set the maximum health of players")
+                .withRequiredArg()
+                .withValuesConvertedBy(new DoubleConverter()
+                        .setPredicate(DoublePredicates.GREATER_THAN_ZERO)
+                        .setType("New maximum player health")
+                );
```
Comment (Eluinhost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java:97]: I like to always give each option a short flag as an alternative to the long flag. Maybe -m and -s for these two? If you could add some example usages of the command to the doc.md that would be good too. `/h -am 40 -s 40` wouldn't be too hard to remember for 20 hearts games. I'm still not 100% sure on the mixing of /h with a set health mechanic tbh but it seems like it should work OK

File: src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java (hunk 1 above)
Comment (ghost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java]: How did I forget about that one. Fixing this after my game

File: src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java (hunk 2 above)
Comment (ghost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java:97]: -m and -s sounds indeed better

**Comment by ghost** [comment]
The client doesn't hard crash on high values, but extremely high values cause FPS lag and will eventually lead to a softlock. 1024 seems like a good hardcoded limit (it's in the default spigot.yml too).

As for the .health permission, [it was already like that in the plugin.yml](https://github.com/Eluinhost/UHC/blob/master/src/main/resources/plugin.yml#L99), so I figured I might aswell fix the documentation (probably should've made it a seperate commit, though).

**Comment by Eluinhost** [comment]
What kind of values does FPS lag normally start for people? Could be worth adding that 'warning' feedback if they set values past it.

If spigot already has a hardcoded limit for max health then what happens if we try and set the health past it? What happens if I run `/h -m2000`? If it auto-limits it then we'll want to also include the max-health in the feedback message so the executor knows the new values. Talking of feedback, should we also be telling the player/s that are being modified their new values like for /heal e.t.c.?

If it was .health already I'll just add a fix message in the changelog, good catch

**Comment by ghost** [comment]
I personally start to lag around ~20000 maximum health (10000 hearts), but this depends on the performance of the player's graphic card.

If a plugin attempts to set the health higher than Spigot's maximum value then it just rolls back to the maximum health. Feedback for the sender is already given as the same logic of showing player health is executed, just after health modification. Telling the player about his health being modified isn't a bad idea though, I'll add that now.

File: src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java [hunk 3]
```diff
@@ -78,15 +78,15 @@ public HealthCommand(MessageTemplates messages, double maximumHealthLimit) {
                 .acceptsAll(ImmutableList.of("a", "all"), "Check/modify all online players");
 
         setCurrentHealthSpec = parser
-                .accepts("set", "Set the health of players")
+                .accepts("s", "Set the health of players")
```
Comment (Eluinhost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java]: You've misunderstood here, you can use both a short and long flag, check out the other commands. `ImmutableSet.of('s', 'set')` + max

**Comment by Eluinhost** [comment]
I meant feedback for the new max health. If I run `/h -am 2000` I'd assume it was successful but spigot would have actually set it to 1024. I'd put the soft limit warning message at > 100 hearts tbh

File: src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java (hunk 3 above)
Comment (ghost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java]: I'm an idiot as usual, yeah I completely forgot that's a thing. Sorry

**Comment by ghost** [comment]
Oh, you mean a warning as in "Expected maximum health to be set to XX but found YY"? That's a good idea.

**Comment by Eluinhost** [comment]
Yeah that would be good, we'd need two new strings:

`Max health was larger than spigot allows, set to {{ amount }} instead`  
`Warning: max health is > 100 hearts, be careful this may start to cause problems for some clients`  

or something similar

File: docs/commands/h.md [hunk 4]
```diff
@@ -1,7 +1,20 @@
...
+`-c`/`-current` - Instead of showing player healths, set their current health instead. Requires permission `uhc.command.health.modify`.
+`-m`/`-maximum` - Instead of showing player healths, set their maximum health instead. Requires permission `uhc.command.health.modify`.
+`-s`/`-silent` - Do not send feedback to players whose health was modified
+
+## Examples
+
+`/h ghowden` - Show health of player named "ghowden"
+`/h -a` - Show all online player's health
+`/h -am 40` - Set maximum health of all players to 40 (20 hearts)
+`/h -sam 40` - Set maximum health of all players to 40 (20 hearts) without sending feedback
```
Comment (Eluinhost) [inline docs/commands/h.md]: Can we switch this example to `/h -sc 60 ghowden Eluinhost` so there is more variation of examples

Comment (ghost) [inline docs/commands/h.md]: Done

**Comment by ghost** [comment]
I think I implemented everything. Anything else?

File: src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java [hunk 5]
```diff
@@ -48,39 +52,133 @@
...
+
+        // Set maximum health before current health in case of multiple operation
+        if (isSettingMaximumHealth) {
+            double newMaximumHealth = setMaximumHealthSpec.value(optionSet);
+
+            boolean limitedByServer = false;
+
+            for (Player player : players) {
+                player.setMaxHealth(newMaximumHealth);
+                double actualMaxHealth = player.getMaxHealth();
```
Comment (Eluinhost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java:134]: No need to call getMaxHealth every iteration when it isn't used

File: src/main/resources/messages.reference.conf [hunk 6]
```diff
@@ -87,15 +87,22 @@
...
       // players = list of players and their health
       "message" : ${colours.command}"Player healths ({{total}}): "${colours.secondary}"{{players}}",
+      "modify message" : ${colours.command}"New player healths ({{total}}): "${colours.secondary}"{{players}}",
       // name = the name of the player
       // percentage = 0-100 percentage health
       // health = 0-20 health
       // maxHealth = player's max health
       "player" : "{{name}} {{percentage}}%",
       "player separator" : " "
+      "modify feedback" : ${colours.command}"Your health has been modified to {{percentage}}%"
```
Comment (Eluinhost) [inline src/main/resources/messages.reference.conf]: Can {{percentage}} here be changed to {{health}}/{{maxHealth}}? Percentage is a bit weird in a notification and excludes any max health changes. Also copy accross the available variables for the message

File: src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java (hunk 5 above)
Comment (ghost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java:134]: It's used on line [135](https://github.com/Eluinhost/UHC/pull/76/files#diff-0794c130ef5bcaad90491b627b9b80b3R135) and [140](https://github.com/Eluinhost/UHC/pull/76/files#diff-0794c130ef5bcaad90491b627b9b80b3R140) though...

File: src/main/resources/messages.reference.conf (hunk 6 above)
Comment (ghost) [inline src/main/resources/messages.reference.conf]: Done

File: src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java (hunk 5 above)
Comment (Eluinhost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java:134]: Fair enough, is there really no better way to detect the max max health?

Comment (Eluinhost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java:134]: Does the value show up in `Server.Spigot.html#getConfig()`?

Comment (ghost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java:134]: Not really, if the maximum health is limited by the server setting  and checking if it actually got set to the expected value is probably the best way. Hooking into Spigot configuration itself sounds like a very nasty hack. (EDIT: And might break future compatibility if Spigot ever decides to change how that logic is handled)

Comment (Eluinhost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java:134]: Fair enough, anything else you want to modify?

Comment (ghost) [inline src/main/java/gg/uhc/uhc/modules/health/HealthCommand.java:134]: Everything seems good.

**Pull request merged**